File: app/src/main/java/com/seafile/seadroid2/framework/model/sdoc/SDocPageOptionsModel.java
package com.seafile.seadroid2.framework.model.sdoc;

import android.os.Parcel;
import android.os.Parcelable;
import android.text.TextUtils;

public class SDocPageOptionsModel implements Parcelable {
    public String docName;
    public String docUuid;
    public String seadocServerUrl;
    public String seadocAccessToken;
    public String repoID;
    public String repoName;
    public boolean isLocked;
    public boolean isStarred;

    public boolean enableMetadataManagement;
    public String latestContributor;
    public String latestContributorName;
    public String lastModificationTime;

    public boolean canUse() {
        if (TextUtils.isEmpty(seadocServerUrl)
                || TextUtils.isEmpty(docUuid)
                || TextUtils.isEmpty(repoID)
                || TextUtils.isEmpty(seadocAccessToken)) {
            return false;
        }

        return true;
    }

    @Override
    public String toString() {
        return "SDocPageOptionsModel{" +
                "docName='" + docName + '\'' +
                ", docUuid='" + docUuid + '\'' +
                ", seadocServerUrl='" + seadocServerUrl + '\'' +
                ", seadocAccessToken='" + seadocAccessToken + '\'' +
                ", repoID='" + repoID + '\'' +
                ", repoName='" + repoName + '\'' +
                ", isLocked=" + isLocked +
                ", isStarred=" + isStarred +
                ", enableMetadataManagement=" + enableMetadataManagement +
                ", latestContributor='" + latestContributor + '\'' +
                ", latestContributorName='" + latestContributorName + '\'' +
                ", lastModificationTime='" + lastModificationTime + '\'' +
                '}';
    }

    @Override
    public int describeContents() {
        return 0;
    }

    @Override
    public void writeToParcel(Parcel dest, int flags) {
        dest.writeString(this.docName);
        dest.writeString(this.docUuid);
        dest.writeString(this.seadocServerUrl);
        dest.writeString(this.seadocAccessToken);
        dest.writeString(this.repoID);
        dest.writeString(this.repoName);
        dest.writeByte(this.isLocked ? (byte) 1 : (byte) 0);
        dest.writeByte(this.isStarred ? (byte) 1 : (byte) 0);
        dest.writeByte(this.enableMetadataManagement ? (byte) 1 : (byte) 0);
        dest.writeString(this.latestContributor);
        dest.writeString(this.latestContributorName);
        dest.writeString(this.lastModificationTime);
    }

    public SDocPageOptionsModel() {
    }

    protected SDocPageOptionsModel(Parcel in) {
        this.docName = in.readString();
        this.docUuid = in.readString();
        this.seadocServerUrl = in.readString();
        this.seadocAccessToken = in.readString();
        this.repoID = in.readString();
        this.repoName = in.readString();
        this.isLocked = in.readByte() != 0;
        this.isStarred = in.readByte() != 0;
        this.enableMetadataManagement = in.readByte() != 0;
        this.latestContributor = in.readString();
        this.latestContributorName = in.readString();
        this.lastModificationTime = in.readString();
    }

    public static final Creator<SDocPageOptionsModel> CREATOR = new Creator<SDocPageOptionsModel>() {
        @Override
        public SDocPageOptionsModel createFromParcel(Parcel source) {
            return new SDocPageOptionsModel(source);
        }

        @Override
        public SDocPageOptionsModel[] newArray(int size) {
            return new SDocPageOptionsModel[size];
        }
    };
}

File: app/src/main/java/com/seafile/seadroid2/framework/model/sdoc/SDocPageOptionsModel.kt
package com.seafile.seadroid2.framework.model.sdoc

import android.os.Parcel
import android.os.Parcelable
import android.text.TextUtils

class SDocPageOptionsModel() : Parcelable {
    @JvmField
    var docName: String? = null

    @JvmField
    var docUuid: String? = null

    @JvmField
    var seadocServerUrl: String? = null

    @JvmField
    var seadocAccessToken: String? = null

    @JvmField
    var repoID: String? = null

    @JvmField
    var repoName: String? = null

    @JvmField
    var isLocked: Boolean = false

    @JvmField
    var isStarred: Boolean = false

    @JvmField
    var enableMetadataManagement: Boolean = false

    fun canUse(): Boolean {
        return !(TextUtils.isEmpty(seadocServerUrl)
            || TextUtils.isEmpty(docUuid)
            || TextUtils.isEmpty(repoID)
            || TextUtils.isEmpty(seadocAccessToken))
    }

    override fun toString(): String {
        return "SDocPageOptionsModel(docName=$docName, docUuid=$docUuid, seadocServerUrl=$seadocServerUrl, seadocAccessToken=$seadocAccessToken, repoID=$repoID, repoName=$repoName, isLocked=$isLocked, isStarred=$isStarred)"
    }

    override fun describeContents(): Int = 0

    override fun writeToParcel(dest: Parcel, flags: Int) {
        dest.writeString(docName)
        dest.writeString(docUuid)
        dest.writeString(seadocServerUrl)
        dest.writeString(seadocAccessToken)
        dest.writeString(repoID)
        dest.writeString(repoName)
        dest.writeByte((if (isLocked) 1 else 0).toByte())
        dest.writeByte((if (isStarred) 1 else 0).toByte())
        dest.writeByte((if (enableMetadataManagement) 1 else 0).toByte())
    }

    private constructor(parcel: Parcel) : this() {
        docName = parcel.readString()
        docUuid = parcel.readString()
        seadocServerUrl = parcel.readString()
        seadocAccessToken = parcel.readString()
        repoID = parcel.readString()
        repoName = parcel.readString()
        isLocked = parcel.readByte().toInt() != 0
        isStarred = parcel.readByte().toInt() != 0
        enableMetadataManagement = parcel.readByte().toInt() != 0
    }

    companion object {
        @JvmField
        val CREATOR: Parcelable.Creator<SDocPageOptionsModel> =
            object : Parcelable.Creator<SDocPageOptionsModel> {
                override fun createFromParcel(source: Parcel): SDocPageOptionsModel =
                    SDocPageOptionsModel(source)

                override fun newArray(size: Int): Array<SDocPageOptionsModel?> = arrayOfNulls(size)
            }
    }
}
